samples/account-linking/csharp/README.md

```diff
@@ -21,9 +23,103 @@ The code generically handles account linking for OAuth2.0, the only GitHub speci
+After the user installs the app, they will be prompted when using any of the capabilities (tab / conversational bot / messaging extension) to log into GitHub. After logging in the app will store the users access and refresh tokens so that they won't need to log in again across the other capabilities or on other devices. 
+
+There are both in-memory and Azure example implementations of the token persistence. 
+    
+You will find the implementation details below. 
+
+This sample will help you
+* Show you the architecture to build out more seamless authentication by using SSO in Teams
+* Use this code (if you build in C#) 
+
+### Scope
+SSO is a problem across multiple surface areas and this sample can be used to build solutions in three surface areas:
+* Tabs
+* Bots
+* Messaging extensions
+
+This sample does not cover:
+* Microsoft Graph tokens. These tokens are received by exchanging user access tokens on the server-side.
+* All languages, it is only built in C#
+* Specific auth systems on your (partner) side
+
+### Context
+#### Problem Overview
+AAD SSO is important for three parties:
+* End user: Reduces the number of sign ins and allows for a single sign-in for all devices
+* Developer: Simpler development when it comes to authentication
+* IT Admin: Protect user privacy and security through conditional access policies
+Often, you (a partner) have your own Identity and Access Management systems and your own user account information. For you to build out AAD SSO inside of Teams, you need to associate the users’ accounts with AAD SSO accounts. 
+
+#### SSO Overview
+Teams enacts AAD SSO in the following manner:
+1.  In the tab, call getAuthToken() using the Teams JavaScript SDK. This tells Teams to obtain an auth token.
+2.  If this is the first time the current user has used your tab application, the user will be prompted to consent.
+3.  Microsoft Teams requests a token for the tab application from the Azure AD endpoint for the current user.
+4.  Azure AD sends the tab application token back to Teams.
+5.  Microsoft Teams then sends the token back to the tab application.
+6.  JavaScript in the tab application can parse the token and extract the information it needs, such as the user's email address. The tab app can optionally exchange the token server-side for further Graph permissions.
+
+### Problem Definition
+Teams SSO (https://docs.microsoft.com/en-us/microsoftteams/platform/tabs/how-to/authentication/auth-aad-sso?tabs=dotnet) offers a seamless experience for partners to build auth inside of Teams. However, Teams SSO is only available for partners who use AzureAD already in their app. For partners that don't use AAD, they often have to use a workaround. There are two main types of scenarios that we typically see partners face when doing these workarounds:
+1.  the user exists in the third-party app with the same email address
+2.  the user exists in the third-party app with a different email address
```

Comment:
I think we're still missing the _major_ reason why these email-centric workarounds don't work.

The core issue isn't establishing an authentic _identity_ for the user (get their email) so much as getting permissions on their behalf. Many partners have outsourced their auth needs to systems like AzureAD B2C, Amazon Cognito, Auth0 or other "whitelabel" identity systems. Their entire systems are anchoring their trust model in those systems and there are two options for dealing with outside identity
1. Side step the existing trust model by creating a carve-out for identity asserted by M365/Azure AD
2. Plug into the ecosystem they've adopted and piggy back on that

Option 1 always results in painful conversations and often deadlocks with security teams and potentially huge architectural changes to enable scenarios to work outside of the identity framework they've chosen. 

Option 2 is what this sample demonstrates. It works out-of-box with zero onus on GitHub to change its concept of "identity" (no change to security model) and requires no carve-out in the existing APIs to enable the integration.